File: lab4/GraphOptimization.java
import java.util.*;

public class GraphOptimization {
    
    // Класс для представления ребра в графе
    static class Edge {
        int target;    // Целевая вершина
        int weight;    // Вес ребра (расстояние, задержка и т.д.)
        
        Edge(int target, int weight) {
            this.target = target;
            this.weight = weight;
        }
    }
    
    // Реализация алгоритма Дейкстры для поиска кратчайших путей
    public static int[] dijkstra(List<List<Edge>> graph, int start) {
        int n = graph.size();
        int[] distances = new int[n];      // Массив для хранения кратчайших расстояний
        boolean[] visited = new boolean[n]; // Массив для отслеживания посещенных вершин
        
        // Инициализация: все расстояния бесконечны, кроме стартовой вершины
        Arrays.fill(distances, Integer.MAX_VALUE);
        distances[start] = 0;
        
        // Приоритетная очередь для эффективного выбора вершины с минимальным расстоянием
        // Хранит пары [вершина, расстояние]
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[1] - b[1]);
        pq.offer(new int[]{start, 0});
        
        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int node = current[0];
            int dist = current[1];
            
            // Пропускаем если уже обработали эту вершину
            if (visited[node]) continue;
            visited[node] = true;
            
            // Обновляем расстояния до всех соседей текущей вершины
            for (Edge edge : graph.get(node)) {
                int newDist = dist + edge.weight;
                // Если найден более короткий путь - обновляем расстояние
                if (newDist < distances[edge.target]) {
                    distances[edge.target] = newDist;
                    pq.offer(new int[]{edge.target, newDist});
                }
            }
        }
        
        return distances;
    }
    
    // ЗАДАНИЕ 1: Оптимизация транспортной сети между складами
    public static void optimizeTransportNetwork() {
        System.out.println("=== ЗАДАНИЕ 1: ОПТИМИЗАЦИЯ ТРАНСПОРТНОЙ СЕТИ ===");
        System.out.println("Цель: найти кратчайшие пути между складами для минимизации логистических затрат");
        
        // Создаем граф транспортной сети (5 складов)
        List<List<Edge>> transportGraph = new ArrayList<>();
        for (int i = 0; i < 5; i++) {
            transportGraph.add(new ArrayList<>());
        }
        
        // Определяем расстояния между складами (в километрах)
        // Склад 0 - главный распределительный центр
        transportGraph.get(0).add(new Edge(1, 50));  // Склад 0 -> Склад 1: 50 км
        transportGraph.get(0).add(new Edge(2, 30));  // Склад 0 -> Склад 2: 30 км
        transportGraph.get(1).add(new Edge(3, 40));  // Склад 1 -> Склад 3: 40 км
        transportGraph.get(2).add(new Edge(3, 20));  // Склад 2 -> Склад 3: 20 км
        transportGraph.get(2).add(new Edge(4, 35));  // Склад 2 -> Склад 4: 35 км
        transportGraph.get(3).add(new Edge(4, 25));  // Склад 3 -> Склад 4: 25 км
        
        // Находим кратчайшие пути от главного склада (0) до всех остальных
        int[] transportDistances = dijkstra(transportGraph, 0);
        
        // Выводим результаты
        System.out.println("\nКратчайшие расстояния от главного склада:");
        String[] warehouseNames = {"Главный склад", "Склад Северный", "Склад Восточный", 
                                  "Склад Южный", "Склад Западный"};
        for (int i = 0; i < transportDistances.length; i++) {
            System.out.println(warehouseNames[i] + " (" + i + "): " + 
                             transportDistances[i] + " км");
        }
        
        // Анализ оптимальных маршрутов
        System.out.println("\nАнализ оптимальных маршрутов:");
        System.out.println("Лучший маршрут до Южного склада: Главный → Восточный → Южный");
        System.out.println("Расстояние: " + transportDistances[3] + " км");
    }
    
    // ЗАДАНИЕ 2: Оптимизация сетевой топологии
    public static void optimizeNetworkTopology() {
        System.out.println("\n=== ЗАДАНИЕ 2: ОПТИМИЗАЦИЯ СЕТЕВОЙ ТОПОЛОГИИ ===");
        System.out.println("Цель: найти пути с минимальной задержкой между серверами");
        
        // Создаем граф сетевой топологии (6 серверов)
        List<List<Edge>> networkGraph = new ArrayList<>();
        for (int i = 0; i < 6; i++) {
            networkGraph.add(new ArrayList<>());
        }
        
        // Определяем задержки передачи между серверами (в миллисекундах)
        // Сервер 0 - главный узел сети
        networkGraph.get(0).add(new Edge(1, 10));  // Сервер 0 -> Сервер 1: 10 мс
        networkGraph.get(0).add(new Edge(2, 15));  // Сервер 0 -> Сервер 2: 15 мс
        networkGraph.get(1).add(new Edge(3, 12));  // Сервер 1 -> Сервер 3: 12 мс
        networkGraph.get(1).add(new Edge(4, 8));   // Сервер 1 -> Сервер 4: 8 мс
        networkGraph.get(2).add(new Edge(4, 20));  // Сервер 2 -> Сервер 4: 20 мс
        networkGraph.get(3).add(new Edge(5, 5));   // Сервер 3 -> Сервер 5: 5 мс
        networkGraph.get(4).add(new Edge(5, 6));   // Сервер 4 -> Сервер 5: 6 мс
        
        // Находим минимальные задержки от главного сервера до всех остальных
        int[] networkDistances = dijkstra(networkGraph, 0);
        
        // Выводим результаты
        System.out.println("\nМинимальные задержки от главного сервера:");
        String[] serverNames = {"Главный сервер", "Сервер БД", "Сервер приложений", 
                               "Файловый сервер", "Веб-сервер", "Резервный сервер"};
        for (int i = 0; i < networkDistances.length; i++) {
            System.out.println(serverNames[i] + " (" + i + "): " + 
                             networkDistances[i] + " мс");
        }
        
        // Анализ оптимальных маршрутов
        System.out.println("\nАнализ оптимальных маршрутов:");
        System.out.println("Оптимальный маршрут до резервного сервера: Главный → Сервер БД → Веб-сервер → Резервный");
        System.out.println("Суммарная задержка: " + networkDistances[5] + " мс");
    }
    
    public static void main(String[] args) {
        // Выполняем оба задания последовательно
        optimizeTransportNetwork();
        optimizeNetworkTopology();
        
        System.out.println("\n=== РЕЗЮМЕ ===");
        System.out.println("Оба примера демонстрируют практическое применение алгоритма Дейкстры");
        System.out.println("для решения реальных задач оптимизации в различных областях");
    }
}

File: lab4/GraphOptimization.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
#include <utility>
#include <string>

using namespace std;

// Структура для представления ребра в графе
struct Edge {
    int target;    // Целевая вершина
    int weight;    // Вес ребра
    
    Edge(int t, int w) : target(t), weight(w) {}
};

// Реализация алгоритма Дейкстры для поиска кратчайших путей
vector<int> dijkstra(const vector<vector<Edge>>& graph, int start) {
    int n = graph.size();
    vector<int> distances(n, INT_MAX);    // Вектор кратчайших расстояний
    vector<bool> visited(n, false);       // Вектор посещенных вершин
    
    // Инициализация: расстояние до стартовой вершины = 0
    distances[start] = 0;
    
    // Приоритетная очередь для хранения пар (расстояние, вершина)
    // greater<> обеспечивает минимальную кучу (min-heap)
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    pq.push({0, start});
    
    while (!pq.empty()) {
        int currentDist = pq.top().first;
        int currentNode = pq.top().second;
        pq.pop();
        
        // Пропускаем если вершина уже обработана
        if (visited[currentNode]) continue;
        visited[currentNode] = true;
        
        // Обновляем расстояния до всех соседних вершин
        for (const Edge& edge : graph[currentNode]) {
            int newDist = currentDist + edge.weight;
            // Если найден более короткий путь - обновляем
            if (newDist < distances[edge.target]) {
                distances[edge.target] = newDist;
                pq.push({newDist, edge.target});
            }
        }
    }
    
    return distances;
}

// ЗАДАНИЕ 1: Оптимизация транспортной сети
void optimizeTransportNetwork() {
    cout << "=== ЗАДАНИЕ 1: ОПТИМИЗАЦИЯ ТРАНСПОРТНОЙ СЕТИ ===" << endl;
    cout << "Цель: найти кратчайшие пути между складами для минимизации логистических затрат" << endl;
    
    // Создаем граф транспортной сети (5 складов)
    vector<vector<Edge>> transportGraph(5);
    
    // Добавляем ребра с расстояниями между складами
    transportGraph[0].push_back(Edge(1, 50));  // Склад 0 -> Склад 1: 50 км
    transportGraph[0].push_back(Edge(2, 30));  // Склад 0 -> Склад 2: 30 км
    transportGraph[1].push_back(Edge(3, 40));  // Склад 1 -> Склад 3: 40 км
    transportGraph[2].push_back(Edge(3, 20));  // Склад 2 -> Склад 3: 20 км
    transportGraph[2].push_back(Edge(4, 35));  // Склад 2 -> Склад 4: 35 км
    transportGraph[3].push_back(Edge(4, 25));  // Склад 3 -> Склад 4: 25 км
    
    // Вычисляем кратчайшие пути от главного склада
    vector<int> transportDistances = dijkstra(transportGraph, 0);
    
    // Выводим результаты
    cout << "\nКратчайшие расстояния от главного склада:" << endl;
    vector<string> warehouseNames = {"Главный склад", "Склад Северный", "Склад Восточный", 
                                    "Склад Южный", "Склад Западный"};
    for (int i = 0; i < transportDistances.size(); i++) {
        cout << warehouseNames[i] << " (" << i << "): " << transportDistances[i] << " км" << endl;
    }
    
    // Анализ результатов
    cout << "\nАнализ оптимальных маршрутов:" << endl;
    cout << "Лучший маршрут до Южного склада: Главный → Восточный → Южный" << endl;
    cout << "Расстояние: " << transportDistances[3] << " км" << endl;
}

// ЗАДАНИЕ 2: Оптимизация сетевой топологии
void optimizeNetworkTopology() {
    cout << "\n=== ЗАДАНИЕ 2: ОПТИМИЗАЦИЯ СЕТЕВОЙ ТОПОЛОГИИ ===" << endl;
    cout << "Цель: найти пути с минимальной задержкой между серверами" << endl;
    
    // Создаем граф сетевой топологии (6 серверов)
    vector<vector<Edge>> networkGraph(6);
    
    // Добавляем ребра с задержками передачи
    networkGraph[0].push_back(Edge(1, 10));  // Сервер 0 -> Сервер 1: 10 мс
    networkGraph[0].push_back(Edge(2, 15));  // Сервер 0 -> Сервер 2: 15 мс
    networkGraph[1].push_back(Edge(3, 12));  // Сервер 1 -> Сервер 3: 12 мс
    networkGraph[1].push_back(Edge(4, 8));   // Сервер 1 -> Сервер 4: 8 мс
    networkGraph[2].push_back(Edge(4, 20));  // Сервер 2 -> Сервер 4: 20 мс
    networkGraph[3].push_back(Edge(5, 5));   // Сервер 3 -> Сервер 5: 5 мс
    networkGraph[4].push_back(Edge(5, 6));   // Сервер 4 -> Сервер 5: 6 мс
    
    // Вычисляем минимальные задержки
    vector<int> networkDistances = dijkstra(networkGraph, 0);
    
    // Выводим результаты
    cout << "\nМинимальные задержки от главного сервера:" << endl;
    vector<string> serverNames = {"Главный сервер", "Сервер БД", "Сервер приложений", 
                                 "Файловый сервер", "Веб-сервер", "Резервный сервер"};
    for (int i = 0; i < networkDistances.size(); i++) {
        cout << serverNames[i] << " (" << i << "): " << networkDistances[i] << " мс" << endl;
    }
    
    // Анализ оптимальных маршрутов
    cout << "\nАнализ оптимальных маршрутов:" << endl;
    cout << "Оптимальный маршрут до резервного сервера: Главный → Сервер БД → Веб-сервер → Резервный" << endl;
    cout << "Суммарная задержка: " << networkDistances[5] << " мс" << endl;
}

int main() {
    // Выполняем оба задания
    optimizeTransportNetwork();
    optimizeNetworkTopology();
    
    cout << "\n=== РЕЗЮМЕ ===" << endl;
    cout << "Оба примера демонстрируют практическое применение алгоритма Дейкстры" << endl;
    cout << "для решения реальных задач оптимизации в различных областях" << endl;
    
    return 0;
}